modules/sdk/gradle-plugins-workspace/src/main/java/com/liferay/gradle/plugins/workspace/internal/client/extension/validator/ClientExtensionValidator.java

```diff
@@ -0,0 +1,159 @@
+
+		if (Objects.equals(clientExtension.type, "themeCSS")) {
+			_validateFrontendTokenDefinitionFile(clientExtension, project);
+		}
+	}
+
+	private void _validateFrontendTokenDefinitionFile(
+		ClientExtension clientExtension, Project project) {
+
+		String frontendTokenDefinitionFilePath = null;
+
+		try {
+			Map<String, Object> typeSettings = clientExtension.typeSettings;
+
+			if (typeSettings.containsKey(_FRONTEND_TOKEN_DEFINITION_JSON_KEY)) {
+				frontendTokenDefinitionFilePath = String.valueOf(
+					typeSettings.get(_FRONTEND_TOKEN_DEFINITION_JSON_KEY));
+
+				File file = project.file(frontendTokenDefinitionFilePath);
+
+				if (!file.exists() || !file.isFile()) {
+					throw new GradleException(
+						String.format(
+							"Unable to find file %s",
+							StringUtil.quote(frontendTokenDefinitionFilePath)));
+				}
+
+				JsonSlurper jsonSlurper = new JsonSlurper();
+
+				jsonSlurper.parse(file);
+			}
+		}
+		catch (JsonException jsonException) {
+			throw new GradleException(
+				String.format(
+					"Unable to parse file %s",
+					StringUtil.quote(frontendTokenDefinitionFilePath)));
+		}
+	}
+
+	private void _validateRequiredDirectory(
+			ClientExtension clientExtension, Project project,
+			String requiredDirectoryName)
+		throws GradleException {
+
+		File file = project.file(requiredDirectoryName);
+
+		if (file.isDirectory()) {
+			return;
+		}
+
+		throw new GradleException(
+			String.format(
+				"A %s directory is required for client extension %s with " +
+					"type %s",
+				StringUtil.quote(requiredDirectoryName), clientExtension.id,
+				clientExtension.type));
+	}
+
+	private void _validateRequiredTypeSettingsKeys(
+			ClientExtension clientExtension, String... requiredTypeSettingsKeys)
+		throws GradleException {
+
+		for (String requiredTypeSettingsKey : requiredTypeSettingsKeys) {
+			if (clientExtension.typeSettings.containsKey(
+					requiredTypeSettingsKey)) {
+
+				continue;
+			}
+
+			throw new GradleException(
+				String.format(
+					"Client extension %s with type %s must define the " +
+						"property %s",
+					clientExtension.id, clientExtension.type,
+					StringUtil.quote(requiredTypeSettingsKey)));
+		}
+	}
+
+	private void _validateTypeSettingsValues(
+			ClientExtension clientExtension, String typeSettingsKey,
+			String... validValues)
+		throws GradleException {
+
+		Object typeSettingsValue = clientExtension.typeSettings.get(
+			typeSettingsKey);
+
+		if ((typeSettingsValue == null) ||
+			ArrayUtil.contains(validValues, typeSettingsValue)) {
+
+			return;
+		}
+
+		throw new GradleException(
+			String.format(
+				"Client extension %s has an invalid value %s for the " +
+					"property %s. Valid values are: %s.",
+				clientExtension.id, StringUtil.quote(typeSettingsValue),
+				StringUtil.quote(typeSettingsKey),
+				com.liferay.petra.string.StringUtil.merge(
+					validValues, StringPool.COMMA_AND_SPACE)));
+	}
+
+	private static final String _FRONTEND_TOKEN_DEFINITION_JSON_KEY =
```

Comment:
Hey @algobob could you redo the commits so that creating the ClientExtensionValidator only involves moving the location of existing code, not changing any existing code. I noticed for instance that _FRONTEND_TOKEN_DEFINITION_JSON_KEY is created in this commit. 

In general "copying" commits like this should only copy code they shouldn't be making any changes to the code being copied. The reason for this is that if you make changes while copying it's to hard to tell what changes were actually made.

Thanks.

Reply:
cool, done